### PR TITLE
[gocryptfs] Add gocryptfs package (build without openssl)

### PR DESCRIPTION
**Pull request by plan5**

### CODE REVIEW

File: package/gocryptfs/package
Comment (Eeems) [inline package/gocryptfs/package]:
```suggestion
    # Official build script runs compiled binary to show version.
```

File: package/gocryptfs/package
Comment (Eeems) [inline package/gocryptfs/package]:
```suggestion
_srcver=2.0-beta2
pkgver=$_srcver-1
```

File: package/gocryptfs/package
Comment (Eeems) [inline package/gocryptfs/package]:
```suggestion
    https://github.com/rfjakob/gocryptfs/archive/v$_srcver.zip
```

File: package/gocryptfs/package
Comment (plan5) [reply package/gocryptfs/package]: Can't merge because I already removed the trailing space on my branch before seeing this.
So it should be taken care of :)

File: package/gocryptfs/package
Comment (Eeems) [inline package/gocryptfs/package]:
```suggestion
    sed -i "s/^.\/gocryptfs.*//" "$srcdir"/build.bash
```

**Review by Eeems — CHANGES_REQUESTED** [review]

**Comment by matteodelabre** [comment]
I would suggest using the recently added [`golang:v1.5` toolchain image](https://github.com/toltec-dev/toolchain#images) and removing the `build:golang` dependency to speed up the build process.

**Comment by plan5** [comment]
> I would suggest using the recently added [`golang:v1.5` toolchain image](https://github.com/toltec-dev/toolchain#images) and removing the `build:golang` dependency to speed up the build process.

Ok, good to know that's there.  I've update the PR (hope I'm doing this right...).

Btw. should I rather try to merge a branch or should I merge directly into testing as I did now?

**Comment by plan5** [comment]
I've now also put all $_srcver in quotes to comply with the style guide.

File: package/gocryptfs/package
Comment (Eeems) [inline package/gocryptfs/package:17]:
```suggestion
```

Since you are using the golang image now.

Comment (Eeems) [inline package/gocryptfs/package]: I'm guessing the build uses git?

Comment (plan5) [reply package/gocryptfs/package]: I'm checking if it's still necessary with the golang image

Comment (plan5) [reply package/gocryptfs/package]: It's not. I've removed it.

**Comment by Eeems** [comment]
@plan5 what would be the best way to test this?

**Comment by plan5** [comment]
> @plan5 what would be the best way to test this?

Are you asking for a manual or an automatized solution? 
Can you send me an example test for another package if for automized testing?

For manual testing:
To test the binary, you can just run it and see if it displays the help message.

To test functionality you can set up any directory to be encrypted. e.g.
```
mkdir /tmp/cipher /tmp/plain
echo "testpass"|gocryptfs -init /tmp/cipher
echo "testpass"|gocryptfs /tmp/cipher /tmp/plain
echo "content" > /tmp/plain/secretfile
fusermount -u /tmp/plain
```

/tmp/plain should now be empty and /tmp/cipher should contain an encrypted file.

```
echo "testpass"|gocryptfs /tmp/cipher /tmp/plain
cat /tmp/plain/secretfile
```
This should output "content".

**Comment by Eeems** [comment]
> Are you asking for a manual or an automatized solution?

Manual testing. 

Thanks, I'll try to take a look later today

**Comment by Eeems** [comment]
Looks like we are missing something? (Testing on a rM1)
image

**Comment by plan5** [comment]
> Looks like we are missing something? (Testing on a rM1)
> image

The gocryptfs -init command only takes one argument (the to be encrypted dir):
`gocryptfs -init /tmp/cipher`

**Comment by plan5** [comment]
But also fuse isn't working, hmm. I'll have to check on rM2 later.

**Comment by Eeems** [comment]
> 
> 
> But also fuse isn't working, hmm. I'll have to check on rM2 later.

Looks like the module isn't included on the rM1. We may need to package that as well. See the wireguard package for information on howw to do that.

**Comment by plan5** [comment]
Ok I'll have a look at it

**Comment by plan5** [comment]
I've added a PR #331 to add a fuse module for rM1 and I've added the module as a dependency for rM1.

**Comment by plan5** [comment]
@Eeems this still shows a requested change from 14 days ago that I've addressed in an independent commit.

**Comment by danshick** [comment]
You can resolve the requested change if it has been addressed. Let's merge the fuse module before this lands.

**Comment by plan5** [comment]
So I had accidentally deleted the gocryptfs branch on github and I've reuploaded it from local. It seems that nothing has broken but I thought I'd better mention this. 
I'm recreating the plan5-fusemodule PR because it was based on plan5-gocryptfs branch instead of testing branch. I will mention this PR there.

Edit: I've resolved it with the existing PR instead.

File: package/gocryptfs/package
Comment (Eeems) [inline package/gocryptfs/package]: This will need to be updated to require the fuse module on rM1. @matteodelabre do we have a mechanism to have different install dependencies between reMarkable versions?

**Review by Eeems — CHANGES_REQUESTED** [review]

File: package/gocryptfs/package
Comment (danshick) [reply package/gocryptfs/package]: We could depend on a metapackage like we do for `display`.

Comment (Eeems) [reply package/gocryptfs/package]: @danshick could you add information to the fuse PR with how to accomplish that for @plan5?

Comment (matteodelabre) [reply package/gocryptfs/package]: To make the dependency rM1-only:

```bash
archs=(rm1 rm2)
# …
installdepends_rm1=(fuse-module)
```

Note that this will generate a separate package for each of the two architectures.

Comment (Eeems) [reply package/gocryptfs/package]: @matteodelabre Which is better, two copies of the package or a metapackage?

Comment (matteodelabre) [reply package/gocryptfs/package]: To be precise, the `display` solution is not technically using a metapackage: the `display` package exists, is empty on rM1, and contains rm2fb on rM2. You could do that also for the fuse module (i.e., make it contain the module on rM1 and be empty on rM2). I think that would preferable since it would save space on our server compared to duplicating the gocryptfs package (and that’s about the only meaningful difference between the two approaches).

Comment (danshick) [reply package/gocryptfs/package]: You're right. Was just looking at it now. I think that's the right approach.

Comment (danshick) [reply package/gocryptfs/package]: Added a comment over in the fuse package pointing to the relevant parts of `display`.

Comment (plan5) [reply package/gocryptfs/package]: I've removed the fuse archs section entirely and added a logic to the package section to only install files on rm1. 
The install dependency in gocryptfs is now only pulled for rm1 via installdepends_rm1.
Does this resolve it all?

Comment (danshick) [reply package/gocryptfs/package]: You can actually now make the fuse package a dependency for all architectures. Because the package is empty for rm2, it'll "install" it, but the package is basically a noop.

This means that there's only one gocryptfs package instead of two, which is desirable because the fuse package is actually the one that needs to vary by device.

Edit:

...and sorry for the complicated advice. We are just now introducing these split architecture packages into testing and still deciding how packages should be built in situations like this one.

Comment (plan5) [reply package/gocryptfs/package]: Ok, I've moved it to the general installdepends.

btw I will add the gocryptfs simple-script GUI as a separate package soon.
I'll have to first find an elegant way to launch it before launchers/xochitl via systemd.
I've already asked about this in the Discord/Matrix-Room but if you have a suggestion I'm all ears :)

Comment (Eeems) [reply package/gocryptfs/package]: I think systemd allows units to specify that they need to launch before something else does.

**Comment by Eeems** [comment]
I'm waiting for the fuse PR to merge before approving this. I'd also like for someone with a rM2 to test the install/uninstall flow with the new fuse dependency.

**Comment by plan5** [comment]
I've just locally merged fuse and then gocryptfs into testing and noticed that #318 conflicts with #331 because I've deleted an accidental copy of the gocryptfs recipe that ended up in the fuse branch. ("CONFLICT (modify/delete): package/gocryptfs/package deleted in HEAD and modified in gocryptfs."). 

Should I just leave it at that and you will fix it when the merge happens or can I do something to smooth it out?

**Comment by Eeems** [comment]
We'll have to resolve the merge conflict after one of them gets merged into testing.

**Review by Eeems — DISMISSED** [review]
Works for me with the latest build of fuse on my rM1. It's probably best to merge this after that PR though.

**Comment by Eeems** [comment]
@plan5 you seem to have removed all changes to this PR?

**Comment by plan5** [comment]
> @plan5 you seem to have removed all changes to this PR?

Oof, I just clicked the button to bring the branch up to date with testing. I thought this would omit the changes to the branch. I'll check out what happened tomorrow and revert it m)

**Comment by plan5** [comment]
> @plan5 you seem to have removed all changes to this PR?

Ok I think it's fixed. It was a mixup because I kept the gocryptfs package in my local testing instead of its own branch. Once it's all merged I'll re-fork and keep things tidy ( I hope ;) )

**Comment by Eeems** [comment]
Looks like you accidentally undid changes on the testing branch.

**Comment by plan5** [comment]
> Looks like you accidentally undid changes on the testing branch.

Ah yes that makes sense, with the revert.

I've merged the official testing branch again. I hope it's fixed now.

**Comment by Eeems** [comment]
It is not, using a revert commit will not allow merges to fix the issue. You'll have to go in and manually revert the changes or do some history rewriting to fix it up.

**Comment by plan5** [comment]
> It is not, using a revert commit will not allow merges to fix the issue. You'll have to go in and manually revert the changes or do some history rewriting to fix it up.

Ok, that's why I tried it with merging testing from toltec-dev into testing from plan5. 
But I can see how this reverts some files to previous states, looking at the PR changes now.

Before doing further radical steps: What do you think about this solution?
1. Clone into toltec-dev/toltec testing
2. Clone into plan5/toltec testing (where the gocryptfs package resides)
3. Delete all files (apart from .git/* and packages/gocryptfs/*) from the local plan5 testing branch
4. Copy all files (apart from .git/* and packages/gocryptfs/*) from the local toltec-dev/toltec testing branch into the local plan5 testing branch
5. git add -A&&git commit -m "Align with current upstream"
6. Push to plan5/toltec testing.

Or would this lead to potential further breakage? I can imagine that this would have to be redone before an actual merge in case there were further changes to the testing branch until then. But as I understand it, they would be applied on top of this commit so it wouldn't be a problem.

The more manual method seems to do the same to me (look at the commit changes of the PR and identify all files that are changed but shouldn't be and undo the changes locally, then commit and push).

**Comment by Eeems** [comment]
At this point I'd just suggest recreating this branch with the files in the state you want them to be in.

**Comment by plan5** [comment]
So that would imply creating a new PR or is there a way to keep this one? I assume recreating the branch would break the PR.

I've tried my suggestion and re-added the gocryptfs package. Right now it would merge without changing anyhing apart from the gocryptfs package, btw.

**Comment by Eeems** [comment]
> So that would imply creating a new PR or is there a way to keep this one? I assume recreating the branch would break the PR.

No need to create a new PR, just force push the new branch.

> I've tried my suggestion and re-added the gocryptfs package. Right now it would merge without changing anyhing apart from the gocryptfs package, btw.

By the looks of it this worked, so we can review it as is instead of replacing the branch.

**Review by Eeems — APPROVED** [review]

**Pull request merged**
